Add QThread/Python threading adaptor

DavidFair opened the pull request:
**Description of work.**
Adds a widget that adapts between Python threading and QThreads. This
provides a number of advantages described in the docstring for
associated classes.

The short-version is it gives us (much) better stack traces, allows us
to get rid of some bare catch exceptions and avoids users having to wire
up Qt signals and slots to simply run something off the GUI thread.

Whilst it's not present in this commit, this has been tested with the
ISIS SANS interface to verify the functionality with Qt and also fix up
some API usability problems noticed whilst implementing this

**To test:**
- Ensure unit tests pass
- Further testing will be present in the follow up PR which replaces the ISIS SANS QThread implementation with this widget

Part of #27895 

This does not require release notes because it's an internal change, though if when we convert users of QThread to this we should explicitly add it as Python stack traces improve greatly

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Add QThread/Python threading adaptor

Adds a widget that adapts between Python threading and QThreads. This
provides a number of advantages described in the docstring for
associated classes.

The short-version is it gives us (much) better stack traces, allows us
to get rid of some bare catch exceptions and avoids users having to wire
up Qt signals and slots to simply run something off the GUI thread.

Whilst it's not present in this commit, this has been tested with the
ISIS SANS interface to verify the functionality with Qt and also fix up
some API usability problems noticed whilst implementing this

DavidFair committed:
Add setting for unit test mode to AsyncQtAdaptor

This will manually change the object under test to run synchronously,
rather than having to faff with trying to get threading right and
possibly having race conditions.

We could include some special logic to manually fire the callbacks on
completion, as per the Qt adaptor layer. However, this encourages people
to write big tests. Instead we encourage them in the docs to separate
the testing for the async method and the testing for the callback(s)

clayton-tom commented on qt/python/mantidqt/utils/async_qt_adaptor.py:
```diff
@@ -0,0 +1,180 @@
+    def _wrap_signal(signal: pyqtSignal, slot: Optional[Callable]) -> Optional[Callable]:
+        """
+        Wraps a given signal so that the thread calls emit rather than the target directly.
+        This allows us to complete the slot on the target, rather than the
```
Think you might've forgotten to finish this sentence

clayton-tom reviewed (changes requested):
New unit looks great and the tests pass, just one comment on the docstring

DavidFair committed:
Finish partial doc string

Completes a doc string which was partially filled in. I think I spotted
something else and left a half-formed thought in its place

DavidFair commented on qt/python/mantidqt/utils/async_qt_adaptor.py:
```diff
@@ -0,0 +1,180 @@
+    def _wrap_signal(signal: pyqtSignal, slot: Optional[Callable]) -> Optional[Callable]:
+        """
+        Wraps a given signal so that the thread calls emit rather than the target directly.
+        This allows us to complete the slot on the target, rather than the
```
Good spot, I had got distracted looking into a test instead :+1:

clayton-tom reviewed (dismissed):
All looks good :+1:

martyngigg commented on qt/python/mantidqt/utils/async_qt_adaptor.py:
```diff
@@ -0,0 +1,181 @@
+from functools import wraps
+from typing import Any, Optional, Callable
+
+from PyQt5 import QtCore
```
Can we swap these for `qtpy` imports?

DavidFair committed:
Switch import from PyQt5 to qtpy

Switches from specific version import to generic qtpy. This originally
used PyQt5 as there was an import not available to qtpy during
development.

However, it appears that whichever import it was, is no longer used
anymore

martyngigg commented on qt/python/mantidqt/utils/async_qt_adaptor.py:
```diff
@@ -0,0 +1,181 @@
+from typing import Any, Optional, Callable
+
+from qtpy import QtCore
+from qtpy.QtCore import QObject, pyqtSignal
```
I forget everytime that in `qtpy` the object is called `Signal` and not `pyqtSignal` 🤦🏻 
 ```suggestion
from qtpy.QtCore import QObject, Signal
```

DavidFair replied:
Weirdly this worked for me locally as I tested before pushing :confused:
